File: CPSC 501 Design patterns/Assignment/11/projectGradingAppJava/src/projectGradingAppJava/ProcessGrades.java
package projectGradingAppJava;

//imports
import java.util.ArrayList; 
import java.io.BufferedWriter;
import java.io.BufferedReader;
import java.io.FileWriter;
import java.io.FileReader;

public class ProcessGrades implements IProcessGrades {

	//list of students
	ArrayList<Student> STList = new ArrayList<Student>();

	public void readStudentData(String inputFileName) throws Exception
	{ // will read students into the STList
		FileReader file = new FileReader(inputFileName);
		BufferedReader sr = new BufferedReader(file);
		try
		{
			STList.clear(); // clear the list of students

			String sline = sr.readLine();
			while (sline != null)
			{
				Student st = null; // base class reference
				String[] parts = sline.split("\t");
				if (parts.length == 6) // undergrad student
				{
					if (parts[3].toUpperCase().equals("UNDERGRAD"))
					{
						st = (Student) new UnderGradStudent(Integer.parseInt(parts[0]),
								parts[1],
								parts[2], Integer.parseInt(parts[4]),
								Integer.parseInt(parts[5]));
					}
				}
				if (parts.length == 7) // grad student
				{
					if (parts[3].toUpperCase().equals("GRADUATE"))
					{
						st = (Student) new GradStudent(Integer.parseInt(parts[0]),
								parts[1],
								parts[2], Integer.parseInt(parts[4]),
								Integer.parseInt(parts[5]), parts[6]);
					}
				}
				if (parts.length == 9) // Phd student
				{
					if (parts[3].toUpperCase().equals("PHDCPSC"))
					{
						st = (Student) new PhdStudent(Integer.parseInt(parts[0]),
								parts[1],
								parts[2], Integer.parseInt(parts[5]),
								Integer.parseInt(parts[6]), parts[7], parts[8]);
					}
				}
				if (st != null)
					STList.add(st); // add student to the list
				sline = sr.readLine(); // read next line
			}
		}
		catch (Exception ex)
		{
			throw ex; // if error, send error back to the calling code
		}
		finally
		{
			sr.close();
		}
	}
	
	public void processAndWriteGrades(String outFileName) throws Exception
	{
		FileWriter file = new FileWriter(outFileName);
		BufferedWriter sw = new BufferedWriter(file);
		try
		{
			STList.sort(null);
			for(Student st:STList) {
				String grade = st.computeGrade();
				// polymorphism, correct ComputeGrade
				// will be called depending upon the type of student in st
				sw.write(st.id + "\t" + grade);
				sw.newLine();
			}
			sw.close();
		}
		catch(Exception ex)
		{
			throw ex;
		}
		finally
		{
			sw.close();
		}
	}
}

File: CPSC 501 Design patterns/Assignment/1/Assignment_1_problem_02/GradingApp/ProcessGrades.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GradingApp
{
    internal class ProcessGrades : IProcessGrades
    {
        public List<Student> STList { get; set; } = new List<Student>();

        public void ProcessAndWriteGrades(string outFileName)
        {
            StreamWriter sw = new StreamWriter(outFileName);
            try
            {
                foreach (Student st in STList)
                {
                    string grade = st.ComputeGrade(); 
    
                    sw.WriteLine(st.ID + "\t" + grade);
                }
                sw.Close();
            }
            catch
            {
                throw; //send error to the caller
            }

            finally // close the stream for this io
            {
                sw.Close();
            }

        }

        public void ReadStudentData(string inputFileName)
        {
            try
            {
                STList.Clear(); 
                StreamReader sr = new StreamReader(inputFileName);
                string sline = sr.ReadLine();
                while (sline != null)
                {
                    Student st = null; //to allow polymorphism and calling different classes
                    string[] parts = sline.Split(new char[] { '\t' });
                    //Console.WriteLine("++++++++++PARTS LEN IS: " + parts.Length);
                    if (parts.Length == 6)  //validations for undergrad student
                    {
                        if (parts[3].ToUpper() == "UNDERGRAD")
                        {
                            st = new UnderGradStudent(int.Parse(parts[0]), parts[1],
                            parts[2], int.Parse(parts[4]), int.Parse(parts[5]));
                        }
                    }
                    if (parts.Length == 7) // validations for grad student
                    {
                        if (parts[3].ToUpper() == "GRADUATE")
                        {
                            st = new GradStudent(int.Parse(parts[0]), parts[1],
                            parts[2], int.Parse(parts[4]), int.Parse(parts[5]),
                           parts[6]);
                        }
                    }
                    if (parts.Length == 9) // validations for Phd student
                    {
                        if (parts[3].ToUpper() == "PHDCPSC")
                        {
                            st = new PhdStudent(int.Parse(parts[0]), parts[1],
                            parts[2], int.Parse(parts[5]), int.Parse(parts[6]),
                           parts[7], parts[8]);
                            Console.WriteLine("PHD " + st.ToString());
                        }
                    }
                    if (st != null) //validate student
                        STList.Add(st); // adding student to the list
                    sline = sr.ReadLine(); // tp the next line
                }
            }
            catch
            {
                throw; //send errors to the caller
            }
        }
    }
}
